detectorSegmentations/src/FCCSWHCalPhiRow_k4geo.cpp

```diff
@@ -0,0 +1,674 @@
+        int idx = - ( floor(irow/m_gridSizeRow[layer]) + 1 );
+        // add the index if it is not already there
+        if(std::find(m_cellIndexes[layer].begin(),m_cellIndexes[layer].end(),idx) == m_cellIndexes[layer].end()) m_cellIndexes[layer].push_back(idx);
+        irow++;
+      }
+    }
+
+    // find edges of each cell in the given layer along z axis
+    for(auto idx : m_cellIndexes[layer])
+    {
+      // calculate z-coordinates of the cell edges
+      double z1 = minLayerZ + (idx-1) * m_dz_row * m_gridSizeRow[layer]; // lower edge
+      double z2 = minLayerZ + (idx) * m_dz_row * m_gridSizeRow[layer]; // upper edge
+
+      // for negative-z Endcap, the index is negative (starts from -1!)
+      if(idx < 0)
+      { 
+        z1 = -minLayerZ + (idx) * m_dz_row * m_gridSizeRow[layer]; // lower edge
+        z2 = -minLayerZ + (idx+1) * m_dz_row * m_gridSizeRow[layer]; // upper edge
+      }
+
+      m_cellEdges[layer][idx] = std::make_pair(z1,z2);
+    }
+
+    dd4hep::printout(dd4hep::DEBUG, "FCCSWHCalPhiRow_k4geo","Number of cells in layer %d: %d", layer, m_cellIndexes[layer].size());
+  }
+}
+
+/// create the cell ID based on the position
+CellID FCCSWHCalPhiRow_k4geo::cellID(const Vector3D& /* localPosition */, const Vector3D& globalPosition,
+                          const VolumeID& vID) const {
+
+  // get the row number from volumeID (starts from 0!)
+  int nrow = _decoder->get(vID, "row");
+  // get the layer number from volumeID
+  uint layer = _decoder->get(vID,"layer");
+
+  CellID cID = vID;
+
+  // get the cell index (start from 1!)
+  int idx = floor(nrow/m_gridSizeRow[layer]) + 1;
+
+  // if the hit is in the negative-z part of the Endcap then assign negative index
+  if(m_offsetZ.size() > 1 && globalPosition.z() < 0) idx *= -1;
+
+  _decoder->set(cID, m_rowID, idx);
+  _decoder->set(cID, m_phiID, positionToBin(dd4hep::DDSegmentation::Util::phiFromXYZ(globalPosition), 2 * M_PI / (double)m_phiBins, m_offsetPhi));
+
+  // For endcap, the volume ID comes with "type" field information which would screw up the topo-clustering,
+  // therefore, lets set it to zero, as it is for the cell IDs in the neighbours map.
+  if(m_offsetZ.size() > 1) _decoder->set(cID, "type", 0);
+
+  return cID;
+}
+
+/// determine the azimuthal angle phi based on the cell ID
+double FCCSWHCalPhiRow_k4geo::phi(const CellID& cID) const {
+  CellID phiValue = _decoder->get(cID, m_phiID);
+  return binToPosition(phiValue, 2. * M_PI / (double)m_phiBins, m_offsetPhi);
+}
+
+/// Get the min and max layer indexes for each part of the HCal
+std::vector<std::pair<uint,uint> > FCCSWHCalPhiRow_k4geo::getMinMaxLayerId() const {
+
+   std::vector<std::pair<uint,uint> > minMaxLayerId;
+
+   if(m_radii.empty()) calculateLayerRadii();
+   if(m_radii.empty()) return minMaxLayerId;
+
+
+   std::vector<int> minLayerIdEndcap = {0, 0, 0};
+   std::vector<int> maxLayerIdEndcap = {-1, 0, 0};
+   if(m_offsetZ.size() > 1)
+   {
+     uint Nl = m_numLayers.size()/3;
```

Comment:
again, hardcoded 3 ..

Reply:
fixed. the code is generalized for N-parts endcap